### DISCUSSION TITLE
Share our graphite storage solution Promate

### DISCUSSION THREAD

**Discussion by faceair** [go-graphite/carbonapi#554] [Show and tell]
https://github.com/zhihu/promate
https://victoriametrics.github.io/CaseStudies.html#zhihu

In our experience, promate is a very big performance improvement over whisper & go-carbon! It is also theoretically better than using victoriametrics directly.

@Civil  Do you have a dataset or method for graphite storage performance testing ?

**Comment by Civil** [comment]
Unfortunately not anything very usable. As while I was working in previous company I thought about doing sort of proper load testing (generating random data that follows real name schema, for that I used to work on creating this https://github.com/go-graphite/gorelka/tree/master/cmd/gorelka-benchmark which generated graphite metrics, but they are not following the pattern :( ) and then replaying the access.log with any load testing tool (almost any can do that). But I haven't managed to finish that up and check if it produce anyhow credible results.

Also I won't be surprised that Promate would work better than whatever internal mechanisms carbonapi provides. As one of the main reasons this support appeared in a first place - is ease of use (single binary that just works for simple use cases).

**Comment by deniszh** [comment]
Hi @faceair 

Thanks, looks very interesting!
Could you maybe answer couple of questions about promate architectire?
1. Why "mateinsert" component needed? Victoriamerics can process  carbon metrics directly, right?
2. Why "matecarbon" component needed? carbonapi can connect to Victoriamerics directly too.

Thanks in advance for answers!

**Reply by deniszh** [reply]
@faceair : Ah, I checked your presentation, much clearer now. I tested VM + carbonapi on Graphite metrics, got very impressive results too.

**Reply by faceair** [reply]
I may need to update the project README, and others may have these questions as well.

There are also some details in the comments of the code, such as how we accomplish real-time aggregation
https://github.com/zhihu/promate/blob/v1.0.0/cmd/matecarbon/main.go#L327-L331

Yes, the performance of the VictoriaMetrics is also very amazing to us.

**Comment by deniszh** [comment]
About benchmark: I had idea to adopt scaled up part of devops tsbs as standard for graphite benchmark metrics - https://github.com/timescale/tsbs
